### PR TITLE
add: separate mod for xdp

### PR DESCRIPTION
**Pull request by syogaraj**
XDP-related operations are now added as a separate mod as discussed in #618 

Closes:
#618 
#205 
#206 
#263

### CODE REVIEW

**Comment by danielocfb** [comment]
Will take a look by Wednesday the latest.

File: libbpf-rs/src/xdp.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        };
+        util::parse_ret(ret)
+    }
+
+    /// Detach the XDP program from the interface
+    pub fn detach(&mut self, ifindex: i32) -> Result<()> {
+        let ret = unsafe {
+            libbpf_sys::bpf_xdp_detach(
+                ifindex,
+                XdpFlags::UPDATE_IF_NOEXIST.bits(),
```
Comment (danielocfb) [inline libbpf-rs/src/xdp.rs:76]: Why are we hardcoding the flags here instead of passing them in?

File: libbpf-rs/src/xdp.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            )
+        };
+        util::parse_ret(ret)
+    }
+
+    /// Query to inspect the program
+    pub fn query(&mut self, ifindex: i32, flags: XdpFlags) -> Result<bpf_xdp_query_opts> {
+        let mut opts = self.query_opts;
+        let err = unsafe {
+            libbpf_sys::bpf_xdp_query(ifindex, flags.bits() as i32, (&mut opts) as *mut _)
```
Comment (danielocfb) [inline libbpf-rs/src/xdp.rs:87]: Why is this cast necessary?
```suggestion
            libbpf_sys::bpf_xdp_query(ifindex, flags.bits() as i32, &mut opts)
```

File: libbpf-rs/src/xdp.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    pub fn query_id(&mut self, ifindex: i32, flags: XdpFlags) -> Result<u32> {
+        let mut prog_id: u32 = 0;
+        let err = unsafe {
+            libbpf_sys::bpf_xdp_query_id(ifindex, flags.bits() as i32, (&mut prog_id) as *mut u32)
+        };
+        if err != 0 {
+            Err(Error::from_raw_os_error(errno::errno()))
+        } else {
+            Ok(prog_id)
+        }
```
Comment (danielocfb) [inline libbpf-rs/src/xdp.rs:106]: Why are we looking at `errno` here instead of parsing the error?

File: libbpf-rs/src/xdp.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    /// Attach the XDP program to the given interface to start processing the packets
+    pub fn attach(&mut self, ifindex: i32, flags: XdpFlags) -> Result<()> {
+        let ret = unsafe {
+            libbpf_sys::bpf_xdp_attach(
+                ifindex,
+                self.fd.as_raw_fd(),
+                flags.bits(),
+                &self.attach_opts as *const _,
```
Comment (danielocfb) [inline libbpf-rs/src/xdp.rs:65]: Cast seems unnecessary, no?

File: libbpf-rs/src/xdp.rs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        /// See [`libbpf_sys::XDP_FLAGS_MASK`].
+        const MASK              = libbpf_sys::XDP_FLAGS_MASK as _;
+    }
+
+}
+
+/// Represents a XDP program.
+///
+/// This struct exposes operations to attach, detach and query a XDP program
+#[derive(Clone, Copy, Debug)]
```
Comment (danielocfb) [inline libbpf-rs/src/xdp.rs:38]: I am not convinced that this type should be cloneable, let alone copyable. What is the reasoning?

File: libbpf-rs/src/xdp.rs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            attach_opts: libbpf_sys::bpf_xdp_attach_opts::default(),
+            query_opts: libbpf_sys::bpf_xdp_query_opts::default(),
+        };
+        xdp.attach_opts.sz = size_of::<libbpf_sys::bpf_xdp_attach_opts>() as libbpf_sys::size_t;
+        xdp.query_opts.sz = size_of::<libbpf_sys::bpf_xdp_query_opts>() as libbpf_sys::size_t;
+        xdp
+    }
+
+    /// Attach the XDP program to the given interface to start processing the packets
+    pub fn attach(&mut self, ifindex: i32, flags: XdpFlags) -> Result<()> {
```
Comment (danielocfb) [inline libbpf-rs/src/xdp.rs:59]: Why do all methods have `&mut self` receivers?

**Review by danielocfb — COMMENTED** [review]
Thanks for your pull request! Your change seems mostly fine. I left a few comments. I think we should think about adding more tests. Specifically, for
- query()
- forced errors for attach(), detach(), query() and query_id()
- double attach() & detach() (second one should fail, I would expect)?

What do you think?

File: libbpf-rs/src/xdp.rs (hunk 4 above)
Comment (syogaraj) [reply libbpf-rs/src/xdp.rs:65]: I took references from `tc.rs` and thought that it was the preferred style of passing opts.

**Comment by syogaraj** [comment]
@danielocfb, 
1. test cases for `query` and `query_id` will not return any error even if the prog is not loaded to the interface. If that's the case, the returned `opts` struct will have the `prog_id` as 0.
2. Added a test case for forced error in the `attach` API. i.e., attach twice and expect an error. 
3. Double detach doesn't seem to raise an error. 
4. Some of the requested changes were made since I took references from `tc.rs` (ex: unnecessary casts and usage of errno). Have updated with the required changes..

**Review by danielocfb — COMMENTED** [review]
Thanks for updating! I left a few more comments regarding the API. Looking closer at it I think there are some cases that we want to support but do not currently. Let me know what you think.

**Comment by danielocfb** [comment]
> 1. test cases for `query` and `query_id` will not return any error even if the prog is not loaded to the interface. If that's the case, the returned `opts` struct will have the `prog_id` as 0.

I think error handling for these functions is broken currently. Don't know if that will change much, though.

File: libbpf-rs/tests/test_xdp.rs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        .query_id(LO_IFINDEX, XdpFlags::UPDATE_IF_NOEXIST)
+        .is_ok());
+
+    assert!(xdp_prog
+        .query(LO_IFINDEX, XdpFlags::UPDATE_IF_NOEXIST)
+        .is_ok());
+
+    assert!(xdp_prog
+        .detach(LO_IFINDEX, XdpFlags::UPDATE_IF_NOEXIST)
+        .is_ok());
```
[the remark this answers is not in the document]
Comment (syogaraj) [reply libbpf-rs/tests/test_xdp.rs:39]: Is there any teardown-like functionality that we can make use of for this? 
Another way I can think of is, having the two tests run in serial while the latter will always be a dummy to call the detach API. What do you think of this?

File: libbpf-rs/src/xdp.rs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+///
+/// This struct exposes operations to attach, detach and query a XDP program
+#[derive(Debug)]
+pub struct Xdp<'fd> {
+    fd: BorrowedFd<'fd>,
+    attach_opts: libbpf_sys::bpf_xdp_attach_opts,
+    query_opts: libbpf_sys::bpf_xdp_query_opts,
+}
+
+impl<'fd> Xdp<'fd> {
```
[the remark this answers is not in the document]
Comment (syogaraj) [reply libbpf-rs/src/xdp.rs:40]: Should we really support flags other than the defined ones? If its user exposed, the flags usage may not be defined. If it's really needed, I can make this change. Please let me know about this.

[the remark this answers is not in the document]
Comment (danielocfb) [reply libbpf-rs/src/xdp.rs:40]: We should not support arbitrary flags. But we should support no flags if that's valid usage from the `libbpf` point of view. `MapFlags` [supports this use case just fine](https://github.com/libbpf/libbpf-rs/blob/9a96b1ef3a98232e15bb7b5d96ed0fd7b37e8a49/libbpf-rs/src/map.rs#L913-L914), for example.

File: libbpf-rs/tests/test_xdp.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (danielocfb) [reply libbpf-rs/tests/test_xdp.rs:39]: > Another way I can think of is, having the two tests run in serial while the latter will always be a dummy to call the detach API. What do you think of this?

I don't think there are any ordering guarantees. Other than mutual exclusion all best are off.

> Is there any teardown-like functionality that we can make use of for this?

I'd create a guard object that detaches on `drop`. Alternatively we could use [`defer`](https://crates.io/crates/defer) for that.

File: libbpf-rs/src/xdp.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (syogaraj) [reply libbpf-rs/src/xdp.rs:40]: Got it. I've added a custom flag `ANY` with the value as `0` to support this and linked this discussion in the comment for future reference.

File: libbpf-rs/tests/test_xdp.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (syogaraj) [reply libbpf-rs/tests/test_xdp.rs:44]: `scopeguard` does provide `defer` macro for this. Have added it in the `test_xdp.rs`

**Review by danielocfb — APPROVED** [review]
Seems okay. Thanks.

**Comment by syogaraj** [comment]
@danielocfb Can you please release a build for this?

**Comment by danielocfb** [comment]
Hi @syogaraj , we just released `0.22.0` and haven't accumulated much since then. Could it wait another week or so?

**Comment by syogaraj** [comment]
We really needs this support in the libbpf-rs for our use case. Can we expect the release in a week? Asking to avoid manually building it from source and use them.

**Comment by danielocfb** [comment]
We can publish a new release this week.

**Comment by danielocfb** [comment]
FYI, I just pushed 0.22.1.

**Pull request merged**